### PR TITLE
feat: add ability to override router addresses in CalculationsSushisw…

### PR DESCRIPTION
**Pull request by x48-crypto**
…ap and add yvBOOST calculations

### CODE REVIEW

File: contracts/Oracle/Calculations/Sushiswap.sol [hunk 1]
```diff
@@ -38,34 +40,28 @@ interface IERC20 {
     function balanceOf(address account) external view returns (uint256);
 }
 
-contract CalculationsSushiswap {
-    address public primaryRouterAddress;
-    address public primaryFactoryAddress;
-    address public secondaryRouterAddress;
-    address public secondaryFactoryAddress;
```
Comment (x48-crypto) [inline contracts/Oracle/Calculations/Sushiswap.sol:45]: Remove concept of primary/secondary router in favor of allowing explicit overrides instead

File: contracts/Oracle/Calculations/Sushiswap.sol [hunk 2]
```diff
@@ -112,18 +108,12 @@ contract CalculationsSushiswap {
...
-                uint256[] memory _amountsOut
-            ) {
-                amountsOut = _amountsOut;
-            } catch {
-                amountsOut = secondaryRouter.getAmountsOut(amountIn, path);
-            }
-        } else {
-            amountsOut = primaryRouter.getAmountsOut(amountIn, path);
+        PriceRouter router = defaultRouter;
+        address routerOverrideAddress = routerOverrides[token0Address];
```
Comment (x48-crypto) [inline contracts/Oracle/Calculations/Sushiswap.sol:112]: Support overrides

File: contracts/Oracle/Calculations/Sushiswap.sol [hunk 3]
```diff
@@ -152,21 +142,6 @@ contract CalculationsSushiswap {
         }
     }
 
-    function getRouterForLpToken(address tokenAddress)
```
Comment (x48-crypto) [inline contracts/Oracle/Calculations/Sushiswap.sol:155]: Remove unused legacy method that relies on specifying factory address. This method is not used in all of lens

File: contracts/Oracle/Calculations/YvBoost.sol [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    constructor(address _addressesProviderAddress) {
+        addressesProviderAddress = _addressesProviderAddress;
+    }
+
+    function setAddressesProviderAddress(address _addressesProviderAddress) public onlyOwner {
+        addressesProviderAddress = _addressesProviderAddress;
+    }
+
+    function getPriceUsdc(address tokenAddress) public view returns (uint256) {
```
Comment (x48-crypto) [inline contracts/Oracle/Calculations/YvBoost.sol:30]: If token is yvBOOST, use sushi calcs instead

File: contracts/Oracle/Calculations/Overrides.sol [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    function setAddressesProviderAddress(address _addressesProviderAddress) public onlyOwner {
+        addressesProviderAddress = _addressesProviderAddress;
+    }
+
+    function getPriceUsdc(address tokenAddress) public view returns (uint256) {
+        string memory overrideId = calculationOverrides[tokenAddress];
+        bool overrideSet = bytes(overrideId).length > 0;
+        if (overrideSet) {
```
Comment (x48-crypto) [inline contracts/Oracle/Calculations/Overrides.sol:37]: Supports arbitrary calculation overrides per asset. This is helpful for yvBOOST, where it's better for us to use CalculationsSushiswap as opposed to CalculationsYearnVaults.

File: contracts/Utilities/Ownable.sol [hunk 6]
```diff
@@ -6,14 +6,18 @@ pragma solidity ^0.8.2;
...
+        ownerAddress = msg.sender;
     }
 
     modifier onlyOwner() {
-        require(msg.sender == owner, "Ownable: caller is not the owner");
+        require(msg.sender == ownerAddress, "Ownable: caller is not the owner");
         _;
     }
+
+    function setOwnerAddress(address _ownerAddress) public onlyOwner {
```
Comment (x48-crypto) [inline contracts/Utilities/Ownable.sol:20]: Add ability to set owner

File: tests/conftest.py [hunk 7]
```diff
@@ -260,12 +258,15 @@ def oracle(
...
-
+    
+    calculationsYearnVaults = CalculationsYearnVaults.deploy(oracle, {"from": management})
     calculationsIronBank = CalculationsIronBank.deploy(
         unitrollerAddress, oracle, {"from": management}
     )
     oracle.setCalculations(
         [
+            calculationsOverrides,
+            calculationsYearnVaults,
```
Comment (x48-crypto) [inline tests/conftest.py:269]: Add calculationsYearnVaults to calculation path before this was missing previously

File: tests/oracle/test_oracle.py [hunk 8]
```diff
@@ -199,18 +216,6 @@ def test_get_price_from_router(oracleProxySushiswap):
     usdcPriceInWeth = oracleProxySushiswap.getPriceFromRouter(usdcAddress, wethAddress)
     assert usdcPriceInEth == usdcPriceInWeth
 
-def test_get_router_for_lp_token(oracleProxySushiswap):
```
Comment (x48-crypto) [inline tests/oracle/test_oracle.py:202]: Remove legacy method that is not used in all of lens (this method requires setting of factory addresses, which is no longer a concept)

File: tests/conftest.py [hunk 9]
```diff
@@ -244,6 +240,8 @@ def oracle(
     usdpAddress = "0x1456688345527bE1f37E9e627DA0837D6f08C925"
     oBtcAddress = "0x8064d9Ae6cDf087b1bcd5BDf3531bD5d8C537a68"
     wbtcAddress = "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599"
+    
+    addressesProviderAddress = "0x9be19Ee7Bc4099D62737a7255f5c227fBcd6dB93"
```
Comment (jstashh) [inline tests/conftest.py:244]: this used?

File: tests/conftest.py (hunk 7 above)
Comment (jstashh) [reply tests/conftest.py:269]: On the live oracle calculationsYearnVaults comes after curve, not sure if that makes a difference or not

File: contracts/Oracle/Calculations/Sushiswap.sol (hunk 2 above)
Comment (jstashh) [reply contracts/Oracle/Calculations/Sushiswap.sol:112]: will uni v3 support the PriceRouter interface, or will we need to deploy a wrapper contract around it?

File: contracts/Oracle/Calculations/Sushiswap.sol [hunk 10]
```diff
@@ -112,18 +108,12 @@ contract CalculationsSushiswap {
...
-            amountsOut = primaryRouter.getAmountsOut(amountIn, path);
+        PriceRouter router = defaultRouter;
+        address routerOverrideAddress = routerOverrides[token0Address];
+        if (routerOverrideAddress != zeroAddress) {
+            router = PriceRouter(routerOverrideAddress);
         }
+        amountsOut = router.getAmountsOut(amountIn, path);
 
         // Return raw price (without fees)
         uint256 amountOut = amountsOut[amountsOut.length - 1];
```
Comment (jstashh) [inline contracts/Oracle/Calculations/Sushiswap.sol:119]: do we need to worry about the fee logic being different on the line below for uniswap v3 positions? Or should we just assume it's always 0.3%

File: tests/oracle/test_oracle.py [hunk 11]
```diff
@@ -168,7 +170,22 @@ def test_cvx_crv_pool_price(oracle):
     assert price > 0
 
 
+# Calculations overrides
+def test_calculations_overrides(oracle, calculationsOverrides, management):
+    yvBOOSTPriceBefore = oracle.getPriceUsdcRecommended(yvBOOSTAddress)
+    calculationsOverrides.setOverrideForToken(yvBOOSTAddress, "CALCULATIONS_SUSHISWAP", {"from": management})
```
Comment (jstashh) [inline tests/oracle/test_oracle.py:176]: What calculations contract was yvboost hitting before? the curve one?

**Review by jstashh — COMMENTED** [review]
lgtm, will approve once the conflicts have been fixed

File: contracts/Oracle/Calculations/Sushiswap.sol (hunk 2 above)
Comment (x48-crypto) [reply contracts/Oracle/Calculations/Sushiswap.sol:112]: We would need to deploy a wrapper contract for uni v3

**Pull request merged**